source/NVDAObjects/UIA/__init__.py

```diff
@@ -142,127 +224,203 @@ def _getControlFieldForObject(self, obj):
-				yield x
-			return
-		tempRange=rangeObj.clone()
-		children=rangeObj.getChildren()
-		for index in xrange(children.length):
-			child=children.getElement(index)
-			childObj=UIA(UIAElement=child.buildUpdatedCache(UIAHandler.handler.baseCacheRequest))
-			#Sometimes a child range can contain the same children as its parent (causing an infinite loop)
-			# Example: checkboxes, graphics, in Edge.
-			if childObj==obj:
+				obj=UIA(windowHandle=self.obj.windowHandle,UIAElement=parentElement)
+				field=self._getControlFieldForObject(obj,isEmbedded=(index==0 and not recurseChildren),startOfNode=startOfNode,endOfNode=endOfNode)
+			except LookupError:
+				log.debug("Failed to fetch controlField data for parentElement. Breaking")
 				continue
-			childRange=self.obj.UIATextPattern.rangeFromChild(child)
-			if childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_End,rangeObj,UIAHandler.TextPatternRangeEndpoint_Start)<=0 or childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_Start,rangeObj,UIAHandler.TextPatternRangeEndpoint_End)>=0:
+			if not field:
 				continue
-			if childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_Start,rangeObj,UIAHandler.TextPatternRangeEndpoint_Start)<0:
-				childRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_Start,rangeObj,UIAHandler.TextPatternRangeEndpoint_Start)
-			if childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_End,rangeObj,UIAHandler.TextPatternRangeEndpoint_End)>0:
-				childRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,rangeObj,UIAHandler.TextPatternRangeEndpoint_End)
-			tempRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,childRange,UIAHandler.TextPatternRangeEndpoint_Start)
-			for f in self._getFormatFieldsAndText(tempRange,formatConfig):
-				yield f
-			field=self._getControlFieldForObject(childObj) if childObj else None
-			if field:
-				yield textInfos.FieldCommand("controlStart",field)
-			for x in self._getTextWithFieldsForRange(childObj,childRange,formatConfig):
-				yield x
-			if field:
-				yield textInfos.FieldCommand("controlEnd",None)
-			tempRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_Start,childRange,UIAHandler.TextPatternRangeEndpoint_End)
-		tempRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,rangeObj,UIAHandler.TextPatternRangeEndpoint_End)
-		for f in self._getFormatFieldsAndText(tempRange,formatConfig):
-			yield f
+			parentFields.append(field)
+		log.debug("Done generating controlFields for parents")
+		log.debug("Yielding control starts for parents")
+		for field in reversed(parentFields):
+			yield textInfos.FieldCommand("controlStart",field)
+		log.debug("Done yielding control starts for parents")
+		del parentElements
+		log.debug("Yielding balanced fields for textRange")
+		# Move through the text range, collecting text and recursing into children
+		tempRange=textRange.clone()

```

Comment:
Can we have a comment explaining that tempRange will be moved to fetch the plain text (if any) between the children? This would make it easier to understand what's going on later. Or is tempRange also used for something else that I'm missing?